File: src/funcoes.cpp
#include "funcoes.hpp"

void AlgoritmosSort::limparArq() {
    ofstream outfile("output.txt", ios::trunc); // Abre o arquivo em modo de truncamento para limpar seu conteúdo
    outfile.close();
}


void AlgoritmosSort::swap(int& a, int& b) {
    int temp = a;
    a = b;
    b = temp;
}

void AlgoritmosSort::selectionSort(vector<int>& arr) {

     auto start = chrono::high_resolution_clock::now();

    int n = arr.size();
    for (int i = 0; i < n - 1; ++i) {
        int minIndex = i;
        for (int j = i + 1; j < n; ++j) {
            if (arr[j] < arr[minIndex]) {
                minIndex = j;
            }
        }
        if (minIndex != i) {
            swap(arr[i], arr[minIndex]);
        }
    }

    auto end = chrono::high_resolution_clock::now();
    chrono::duration<double> duration = end - start;

    ofstream outfile("output.txt", ios_base::app); // Abre o arquivo em modo de adição
    outfile << "[ Tempo de execução do Selection Sort ] : [ " << duration.count() << " segundos ]\n";
    outfile.close();
}

void AlgoritmosSort::insertionSort(vector<int>& arr) {

     auto start = chrono::high_resolution_clock::now();


    int n = arr.size();
    for (int i = 1; i < n; ++i) {
        int key = arr[i];
        int j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j = j - 1;
        }
        arr[j + 1] = key;
    }

    auto end = chrono::high_resolution_clock::now();
    chrono::duration<double> duration = end - start;

    ofstream outfile("output.txt", ios_base::app); // Abre o arquivo em modo de adição
    outfile << "[ Tempo de execução do Insertion Sort ] : [ " << duration.count() << " segundos ]\n";
    outfile.close();
}

void AlgoritmosSort::gnomeSort(vector<int>& arr) {

     auto start = chrono::high_resolution_clock::now();

    int n = arr.size();
    int index = 0;
    while (index < n) {
        if (index == 0) {
            index++;
        }
        if (arr[index] >= arr[index - 1]) {
            index++;
        } else {
            swap(arr[index], arr[index - 1]);
            index--;
        }
    }

    auto end = chrono::high_resolution_clock::now();
    chrono::duration<double> duration = end - start;

    ofstream outfile("output.txt", ios_base::app); // Abre o arquivo em modo de adição
    outfile << "[ Tempo de execução do Gnome Sort ] : [ " << duration.count() << " segundos ] \n";
    outfile.close();
}

vector<int> AlgoritmosSort::lerArq(const string& nomeArquivo, int tamanhoVetor) {
    ifstream arquivo(nomeArquivo);
    if (!arquivo.is_open()) {
        cerr << "Erro ao abrir o arquivo." << endl;
        return {};
    }

    vector<int> vetor;
    int numero;
    while (arquivo >> numero && static_cast<int>(vetor.size()) < tamanhoVetor) {
        vetor.push_back(numero);
    }

    arquivo.close();
    return vetor;
}

File: src/funcoes.py
import time
def selection_sort(arr):
    start_time = time.time()
    n = len(arr)
    for i in range(n-1):
        min_index = i
        for j in range(i+1, n):
            if arr[j] < arr[min_index]:
                min_index = j
        if min_index != i:
            arr[i], arr[min_index] = arr[min_index], arr[i]
    end_time = time.time()
    elapsed_time = end_time - start_time
    with open("output.txt", "a") as f:
        f.write("[ Tempo de execução do Selection Sort ] : [ {:.6f} segundos ]\n".format(elapsed_time))

def insertion_sort(arr):
    start_time = time.time()
    n = len(arr)
    for i in range(1, n):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
    end_time = time.time()
    elapsed_time = end_time - start_time

    with open("output.txt", "a") as f:
        f.write("[ Tempo de execução do Insertion Sort ] : [ {:.6f} segundos ]\n".format(elapsed_time))

def gnome_sort(arr):
    start_time = time.time()
    n = len(arr)
    index = 0
    while index < n:
        if index == 0:
            index += 1
        if arr[index] >= arr[index - 1]:
            index += 1
        else:
            arr[index], arr[index - 1] = arr[index - 1], arr[index]
            index -= 1
    end_time = time.time()
    elapsed_time = end_time - start_time

    with open("output.txt", "a") as f:
        f.write("[ Tempo de execução do Gnome Sort ] : [ {:.6f} segundos ]\n".format(elapsed_time))

def lerArq(nome_arquivo, tamanho_vetor):
    try:
        with open(nome_arquivo, 'r') as arquivo:
            numeros = []
            for linha in arquivo:
                if len(numeros) < tamanho_vetor:
                    numero = int(linha.strip())
                    numeros.append(numero)
                else:
                    break
        return numeros
    except FileNotFoundError:
        print("Erro ao abrir o arquivo.")
        return []

def limparArq():
    while True:
        choice = input("Deseja LIMPAR o Arquivo? [S/N]: ").upper()
        if choice == 'S':
            with open("output.txt", "w") as f:
                f.write("")  # Limpa o conteúdo do arquivo
            print("Conteúdo do arquivo limpo.")
            break
        elif choice == 'N':
            print("Nenhuma ação realizada.")
            break
        else:
            print("Opção inválida. Por favor, digite 's' para sim ou 'n' para não.")